patch/cisco-mtd-part.patch

```diff
@@ -0,0 +1,33 @@
+From 157c059a2c639ecc69080ecb004a94e652467a4a Mon Sep 17 00:00:00 2001
+From: Madhava Reddy Siddareddygari <msiddare@cisco.com>
+Date: Tue, 8 Jun 2021 10:19:36 -0700
+Subject: [PATCH] mtd: Add support for reading partition data from ACPI
+
+Current mtd driver supports reading mtd partition information from
+command line and dts file, but cannot read partition information
+from ACPI files.
+
+Added code to support new partition type "acpipart" to
+default_mtd_part_types structure. This partition type can be used
+to configure mtd partition data in ACPI tables.
+
```

Comment:
Tested how?

Grepping the Linux source code for “acpipart” does not give any results. Can you please provide more information, how it works.

Reply:
"acpipart" is used by cisco specific custom driver. 
Please see the code snippet which registers "apcipart" as mtd parser function.

static struct mtd_part_parser cisco_acpi_mtd_part_parser = {
    .parse_fn = cisco_acpi_mtd_part_parse,
    .name = "acpipart",
};

static struct mtd_part_parser cisco_acpi_mtd_part_parser_ofpart = {
    .parse_fn = cisco_acpi_mtd_part_parse,
    .name = "ofpart",
};

static int __init cisco_acpi_mtd_part_init(void)
{
    register_mtd_parser(&cisco_acpi_mtd_part_parser);
    register_mtd_parser(&cisco_acpi_mtd_part_parser_ofpart);
    return 0;
}

Example snippet of configuring mtd partition through ACPI tables for a variants of cisco-8000 platform
Scope ({{ PEMBREY_SCOPE }}.PMBR.SPIA)
{
    Name (_HID, "PRP0001")
    Name (_STR, Unicode ("SPI Aldrin"))
    Name (_DSD, Package ()
    {
        ToUUID ("daffd814-6eba-4d8c-8a91-bc9bbf4aa301"),
        Package () {
            Package () { "devid-auto", 1 },
            Package () { "compatible", "cisco-fpga-spi" },
        }
    })
    Device(ALD0)
    {
        Name (_HID, "ACPI0000")
        Name (_STR, Unicode ("Aldrin flash"))
        Name (_DSD, Package ()
        {
            ToUUID ("daffd814-6eba-4d8c-8a91-bc9bbf4aa301"),
            Package () {
                Package () { "compatible", "jedec,spi-nor" },
                Package () { "fixed-partition-labels", Package () {
                    "aldrin-upgrade-microinit",
                    "aldrin-golden-microinit",
                    "aldrin-metadata",
                }},
                Package () { "fixed-partition-start", Package () {
                    0x0000000,
                    0x0200000,
                    0x1FC0000,
                }},
                Package () { "fixed-partition-size", Package () {
                    0x0200000,
                    0x0200000,
                    0x0040000,
                }},
            }
        })
        Name (_CRS, ResourceTemplate () {
            SPISerialBus(0,                     // device selection
                         PolarityLow,           // device selection polarity
                         FourWireMode,          // wiremode
                         8,                     // databit len
                         ControllerInitiated,   // slave mode
                         2000000,               // connection speed
                         ClockPolarityLow,      // clock polarity
                         ClockPhaseSecond,      // clock phase
                         "^^SPIA",              // RscSource: SPI bus controller
                         0,                     // ResourceSourceIndex
                                                // Resource usage
                                                // Vendor data
            )
        })
    }
}

Output on the target system:
cisco@sonic:~$ cat /proc/mtd 
dev:    size   erasesize  name
mtd0: 01000000 00001000 "BIOS"
mtd1: 00200000 00040000 "aldrin-upgrade-microinit"
mtd2: 00200000 00040000 "aldrin-golden-microinit"
mtd3: 00040000 00040000 "aldrin-metadata"

Reply:
It’d be great, if you marked up the pastes/code blocks by intending them by four spaces or enclosing them with ``` to improve legibility?

Reply:
Sorry, revisiting this, is the custom Cisco driver public?

Reply:
Hi Paul,

Cisco custom driver is not public

Reply:
It needs to be public, doesn’t it? Otherwise SONiC won’t run well on the device?

Reply:
Cisco custom driver is build only when platform is selected as cisco 

make configure PLATFORM=cisco-8000 will set platform as cisco and "make all" will take care of integrating all cisco packages required for Sonic 

Cisco driver which uses "acpipart" is custom and driver source is not shared to public.